core/src/muxing.rs

```diff
@@ -431,8 +439,33 @@ where
     #[inline]
     fn shutdown(&mut self) -> Poll<(), io::Error> {
         let s = self.substream.as_mut().expect("substream was empty");
-        self.muxer.shutdown_substream(s).map_err(|e| e.into())?;
-        Ok(Async::Ready(()))
+        loop {
+            match self.shutdown_state {
+                ShutdownState::Shutdown => {
+                    match self.muxer.shutdown_substream(s).map_err(|e| e.into())? {
```

Comment:
Could you do something like this?

```rust
let _ = try_ready!(self.muxer.shutdown_substream(s));
self.shutdown_state = ShutdownState::Flush;
```

(and same below)

It would make that code much more concise.

Reply:
Done.